File: frontend/src/api/html.ts
import { apiClient } from '@/api/client';
import type {
  HtmlVersionContent,
  HtmlVersionSaveRequest,
  HtmlVersionSaveResponse,
  HtmlPosterOut,
  HtmlGenerateRequest,
  HtmlGenerateResponse,
} from '@/api/types';

export const htmlApi = {
  /** 获取指定版本的 HTML 内容 */
  getVersionContent(versionId: string): Promise<HtmlVersionContent> {
    return apiClient.get<HtmlVersionContent>(`/api/html/${versionId}`);
  },

  /** 获取海报详情及版本列表 */
  getPoster(posterId: string): Promise<HtmlPosterOut> {
    return apiClient.get<HtmlPosterOut>(`/api/html/poster/${posterId}`);
  },

  /** 列出活动下所有 HTML 海报 */
  listPosters(campaignId: string): Promise<HtmlPosterOut[]> {
    return apiClient.get<HtmlPosterOut[]>(`/api/campaigns/${campaignId}/html`);
  },

  /** 保存手动编辑版本（不覆盖旧版本） */
  saveVersion(posterId: string, body: HtmlVersionSaveRequest): Promise<HtmlVersionSaveResponse> {
    return apiClient.post<HtmlVersionSaveResponse>(`/api/html/${posterId}/versions`, body);
  },

  /** 基于当前底图重新生成 HTML */
  regenerateHtml(campaignId: string, body: HtmlGenerateRequest): Promise<HtmlGenerateResponse> {
    return apiClient.post<HtmlGenerateResponse>(`/api/campaigns/${campaignId}/html/generate`, body);
  },

  /** 预览 URL（供 iframe src 使用，返回 text/html） */
  getPreviewUrl(versionId: string): string {
    return `/api/html/${versionId}/preview`;
  },
};

File: backend/app/schemas/html.py
"""HtmlPoster / HtmlVersion schemas（落地方案 §5.6–5.7 / §9.8–9.12）。"""

from datetime import datetime
from typing import Any

from pydantic import BaseModel


# ── HTML 生成 ─────────────────────────────────────────────────────────────────

class HtmlGenerateRequest(BaseModel):
    """POST /api/campaigns/{id}/html/generate 请求体（落地方案 §9.8）。"""

    selected_image_id: str
    model: str | None = None


class ValidationResult(BaseModel):
    ok: bool
    issues: list[str] = []


class HtmlGenerateResponse(BaseModel):
    """POST /api/campaigns/{id}/html/generate 响应体（落地方案 §9.8）。"""

    poster_id: str
    version_id: str
    status: str
    preview_url: str
    validation: ValidationResult


# ── HTML 内容读取 ─────────────────────────────────────────────────────────────

class HtmlVersionContent(BaseModel):
    """GET /api/html/{version_id} 响应体（落地方案 §9.9）。"""

    version_id: str
    poster_id: str
    version_no: int
    source: str
    html: str
    created_at: datetime


# ── 保存手动编辑版本 ──────────────────────────────────────────────────────────

class HtmlVersionSaveRequest(BaseModel):
    """POST /api/html/{poster_id}/versions 请求体（落地方案 §9.10）。"""

    source: str = "manual_edit"
    html: str


class HtmlVersionSaveResponse(BaseModel):
    """POST /api/html/{poster_id}/versions 响应体（落地方案 §9.10）。"""

    poster_id: str
    version_id: str
    version_no: int
    validation: ValidationResult


# ── 版本列表 ──────────────────────────────────────────────────────────────────

class HtmlVersionOut(BaseModel):
    id: str
    poster_id: str
    version_no: int
    source: str
    html_path: str | None = None
    model: str | None = None
    validation: dict[str, Any] | None = None
    created_at: datetime

    model_config = {"from_attributes": True}


class HtmlPosterOut(BaseModel):
    id: str
    campaign_id: str
    selected_image_id: str | None = None
    title: str | None = None
    current_version_id: str | None = None
    status: str
    created_at: datetime
    updated_at: datetime
    versions: list[HtmlVersionOut] = []

    model_config = {"from_attributes": True}
